Imhex darwin

pcasaretto opened the pull request:
## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

pcasaretto commented:
It seems we reached the same block @afh. It's the bundle step that fails when it tries to copy the system libraries and fails to find them.

emilazy commented:
If it’s trying to copy libraries into the bundle, we can probably just skip it. Nix application bundles usually depend on the store anyway.

pcasaretto commented:
Interesting. According to the [docs for `fixup_bundle`](https://cmake.org/cmake/help/latest/module/BundleUtilities.html) we could pass a a list of items to ignore

emilazy commented:
It can just be removed entirely.

pcasaretto commented:
The call to `fixup_bundle` you mean?

emilazy commented:
Yes, it shouldn’t be necessary unless we wanted to distribute a standalone `.app` bundle.

pcasaretto commented:
Well, it seems it is at least partially necessary: https://github.com/WerWolv/ImHex/issues/1807#issuecomment-2225791612

pcasaretto commented:
I might be wrong, that screenshot is from skipping the entire "packaging" step. I'm testing removing only the `fixup_bundle` instruction.

pcasaretto commented:
<img width="1112" alt="image" src="https://github.com/user-attachments/assets/91788806-c2f3-4677-b42c-4a80b0513113">
It seems we do need at least some of it.

pcasaretto commented:
https://github.com/pcasaretto/ImHex/commit/a1bc0806b6b46bc0c240f73f17c38be257ba1d97
resulted in:

```
-- fixup_bundle                                                                                                                                       
--   app='/nix/store/dg3d5zx6flr60xzy5s3mkpi04pwadjh5-imhex-1.35.4/imhex.app'                                                                         
--   libs='/nix/store/dg3d5zx6flr60xzy5s3mkpi04pwadjh5-imhex-1.35.4/imhex.app/Contents/MacOS/plugins/builtin.hexplug;/nix/store/dg3d5zx6flr60xzy5s3mkp
i04pwadjh5-imhex-1.35.4/imhex.app/Contents/MacOS/plugins/decompress.hexplug;/nix/store/dg3d5zx6flr60xzy5s3mkpi04pwadjh5-imhex-1.35.4/imhex.app/Content
s/MacOS/plugins/diffing.hexplug;/nix/store/dg3d5zx6flr60xzy5s3mkpi04pwadjh5-imhex-1.35.4/imhex.app/Contents/MacOS/plugins/disassembler.hexplug;/nix/st
ore/dg3d5zx6flr60xzy5s3mkpi04pwadjh5-imhex-1.35.4/imhex.app/Contents/MacOS/plugins/hashes.hexplug;/nix/store/dg3d5zx6flr60xzy5s3mkpi04pwadjh5-imhex-1.
35.4/imhex.app/Contents/MacOS/plugins/script_loader.hexplug;/nix/store/dg3d5zx6flr60xzy5s3mkpi04pwadjh5-imhex-1.35.4/imhex.app/Contents/MacOS/plugins/
visualizers.hexplug;/nix/store/dg3d5zx6flr60xzy5s3mkpi04pwadjh5-imhex-1.35.4/imhex.app/Contents/MacOS/plugins/yara_rules.hexplug'                     
--   dirs='/var/empty/local/lib;/lib;/var/empty/lib;/nix/store/dg3d5zx6flr60xzy5s3mkpi04pwadjh5-imhex-1.35.4/imhex.app/Contents/MacOS/plugins;/nix/sto
re/dg3d5zx6flr60xzy5s3mkpi04pwadjh5-imhex-1.35.4/imhex.app/Contents/Frameworks'                                                                       
--   ignoreItems='/var/empty/lib/libSystem.B.dylib'                                                                                                   
-- fixup_bundle: preparing...                                                                                                                         
--                                                                                                                                                    
warning: cannot resolve item '/usr/lib/libSystem.B.dylib'                                                                                             
                                                                                                                                                      
  possible problems:                                                                                                                                  
    need more directories?                                                                                                                            
    need to use InstallRequiredSystemLibraries?                                                                                                       
    run in install tree instead of build tree?
```

afh commented:
Thank you for this PR, @pcasaretto, that's very helpful!

I tried add `darwin.apple_sdk_11_0.frameworks.System` to the `buildInputs`, unfortunately that did not resolve the current issue with `otool` and `libSystem`; it seems we need to dig deeper…

While work continues on the libSystem issue a few minor notes on this PR:

* Could you please split the commits into two: one for the darwin specific bits and one for the refactoring changes
* I suggest to make the commit message for the former `imhex: add support for darwin` and for the latter `imhex: refactor`.<sup>1</sup>

─────
<sup>1</sup> Thanks for the kudos in the commit message, very much appreciated 🙏 Although from a project perspective that provides little value to others. In case you wanted to show appreciation for my help with this maybe you'd like to return the favor and have a look at [the work that I'm doing in nixpkgs](https://github.com/NixOS/nixpkgs/pulls/afh) and provide feedback and support? 🙂

afh commented on pkgs/by-name/im/imhex/package.nix:
```diff
@@ -60,29 +62,33 @@ stdenv.mkDerivation rec {
     mkdir -p $out/share/imhex
-    rsync -av --exclude="*_schema.json" ${patterns_src}/{constants,encodings,includes,magic,patterns} $out/share/imhex
+    cp -aR ${patterns_src}/{constants,encodings,includes,magic,patterns} $out/share/imhex
+    find $out/share/imhex -name '*_schema.json' -exec rm {} \;
```
* After building this locally `share/imhex/constants/_schema.json` still exists, I'm uncertain why the find command does not properly remove it 🤔 

* A ~~solution~~workaround for the current darwin build issue is to manually fix the reference to `libimhex` in the app using `install_name_tool`. I'm not particularly fond of doing it this way and maybe someone from @NixOS/darwin-maintainers has a better approach or idea.
```suggestion
    find $out/share/imhex -name '*_schema.json' -exec rm {} \;
  '' + lib.optionalString stdenv.isDarwin ''
    mkdir -p $out/Applications
    mv $out/imhex.app $out/Applications
    install_name_tool \
      -change "$out/lib/libimhex.${finalAttrs.version}${stdenv.hostPlatform.extensions.sharedLibrary}" \
      "@executable_path/../Frameworks/libimhex.${finalAttrs.version}${stdenv.hostPlatform.extensions.
⣀ sharedLibrary}" \
      "$out/Applications/imhex.app/Contents/MacOS/imhex"
```

pcasaretto replied:
What was the motivation for the removal of rsync? Was it necessary?

afh commented on pkgs/by-name/im/imhex/package.nix:
```diff
@@ -60,29 +62,33 @@ stdenv.mkDerivation rec {
 
   meta = with lib; {
-    description = "Hex Editor for Reverse Engineers, Programmers and people who value their retinas when working at 3 AM";
+    description = "Hex Editor for Reverse Engineers, Programmers";
```
I think this reads nicer 🙂 
```suggestion
    description = "Hex Editor for Reverse Engineers and Programmers";
```

pcasaretto replied:
It seems it was copied from the official about
<img width="311" alt="image" src="https://github.com/user-attachments/assets/9a45e0b5-a0ec-490a-ad5c-ae186eaca7c7">
https://github.com/WerWolv/ImHex

afh commented on pkgs/by-name/im/imhex/package.nix:
```diff
@@ -33,19 +35,19 @@ let
   };
 
-  nativeBuildInputs = [ cmake llvm python3 perl pkg-config rsync ];
+  nativeBuildInputs = [ cmake llvmPackages.llvm python3 perl pkg-config ];
```
It is possible to skip CMake's `fixup_bundle` and manually fixup the library paths (see suggestions on the `postInstall` phase below).
```suggestion
  postPatch = lib.optionalString stdenv.isDarwin ''
    substituteInPlace cmake/modules/PostprocessBundle.cmake \
      --replace-fail "fixup_bundle" "#fixup_bundle"
  '';

  nativeBuildInputs = [ cmake llvmPackages.llvm python3 perl pkg-config ];
```

afh reviewed (changes requested):
Please find below a few suggestions that fix the build on `aarch64-darwin` for me and I'm able to build, open, and run imhex 🥳 

I've not tested imhex's functionality, mind taking a closer look to ensure it works as you'd expect, @pcasaretto ?

afh reviewed (changes requested):
For some reason GitHub posted this twice.
~~Please find below a few suggestions that fix the build on `aarch64-darwin` for me and I'm able to build, open, and run imhex 🥳~~

~~I've not tested imhex's functionality, mind taking a closer look to ensure it works as you'd expect?~~

pcasaretto commented:
Thank you for your awesome contributions @afh! I'll work on this later today.
I'll definitely (try and) return the favor.

afh commented:
Thanks for the opportunity to work on this, @pcasaretto, it was fun!

inclyc commented:
Never mind, the bot seems to be misbehaved.

pcasaretto commented:
Successfully built and run. However, it seems that we might be missing some functionality related to patterns.
Open a Gameboy ROM with the pre-built image from the official Releases I get this behavior:

<img width="1392" alt="image" src="https://github.com/user-attachments/assets/4aec1c3b-9c0c-463f-a3dd-a407952925bf">

The same does not happen with our version.
Investigating.

pcasaretto commented:
<img width="988" alt="image" src="https://github.com/user-attachments/assets/77908d64-14d6-4dcf-90c1-ead43b0c2080">

The official version has all these resources packaged in.

Our version is rather barren 😅 

<img width="1032" alt="image" src="https://github.com/user-attachments/assets/43b33fcd-45a3-45c3-a0f3-12e62833e06d">

nixf-tidy-review commented on pkgs/by-name/im/imhex/package.nix:
```diff
@@ -60,29 +67,41 @@
+    find $out/share/imhex -name '*_schema.json' -exec rm {} \;
+  '' + lib.optionalString stdenv.isDarwin ''
+    mkdir -p $out/Applications
+    mv $out/imhex.app $out/Applications
```
extra `.` at the end of attrpath

nixf-tidy-review commented on pkgs/by-name/im/imhex/package.nix:
```diff
@@ -60,29 +67,41 @@
+  '' + lib.optionalString stdenv.isDarwin ''
+    mkdir -p $out/Applications
+    mv $out/imhex.app $out/Applications
+    install_name_tool \
```
unexpected text

nixf-tidy-review replied:
expected }

nixf-tidy-review commented:
There seems to be syntax errors in your revision dfc0609a0b0a68cf7ec40a996d33f96f3e4a1f5d, CC @pcasaretto. If you feel I'm reporting false positives, Please ping <code>@inclyc</code>

pcasaretto commented:
Well, we just need to copy the resources in `postInstall` like it's done for `$out/share`. Working on it.

pcasaretto commented:
Partial success. Still missing some stuff.
I've also taken the opportunity to add a `makeWrapper` call.

```
❯ tree -L 1 /nix/store/ja7j1jbk2frhslfgh934x47j0l0c5gcv-imhex-1.35.4/Applications/imhex.app/Contents/MacOS/
/nix/store/ja7j1jbk2frhslfgh934x47j0l0c5gcv-imhex-1.35.4/Applications/imhex.app/Contents/MacOS/
├── constants
├── encodings
├── imhex
├── imhex-updater
├── includes
├── magic
├── patterns
└── plugins

7 directories, 2 files
❯ tree -L 1 /Volumes/ImHex/ImHex.app/Contents/MacOS/
/Volumes/ImHex/ImHex.app/Contents/MacOS/
├── constants
├── encodings
├── fonts.hexpluglib
├── imhex
├── imhex-updater
├── includes
├── magic
├── nodes
├── patterns
├── plugins
└── ui.hexpluglib

8 directories, 4 files
```

I'm going to investigate a little further but we are really close.

pcasaretto commented:
Using otool I noticed there were some dependencies on

viraptor commented:
I just checked the `fonts.hexpluglib` and `ui.hexpluglib` and they are installed, but are under plugins instead with the `*.hexplug` files.
They get loaded correctly too (removing them break basic functionality). So we don't need to worry about those two.

viraptor commented on pkgs/by-name/im/imhex/package.nix:
```diff
@@ -60,29 +78,41 @@ stdenv.mkDerivation rec {
-  # rsync is used here so we can not copy the _schema.json files
   postInstall = ''
     mkdir -p $out/share/imhex
     rsync -av --exclude="*_schema.json" ${patterns_src}/{constants,encodings,includes,magic,patterns} $out/share/imhex
```
```suggestion
    rsync -av --exclude="*_schema.json" ${patterns_src}/{constants,encodings,includes,magic,nodes,patterns} $out/share/imhex
```

afh replied:
What can we do to avoid use of rsync? It feels unnecessary...

pcasaretto replied:
Great! That matches the answers I got from the author. Thank you. 
Is the share folder relevant on Darwin or should we only do that on Linux?

toonn replied:
`find ... -exec cp {} target/` is always an option. I agree that rsync should not be a dependency for a package build unless its incremental transfers actually provide a speedup/reduce network traffic.

emilazy replied:
IMO `rsync` is fine here, it’s just a souped‐up `cp` in many ways and it doesn’t have many dependencies. No objections to replacing it with `find`, though.

viraptor commented on pkgs/by-name/im/imhex/package.nix:
```diff
@@ -60,29 +78,41 @@ stdenv.mkDerivation rec {
+  '' + lib.optionalString stdenv.isDarwin ''
+    mkdir -p $out/Applications
+    mv $out/imhex.app $out/Applications
+    rsync -av --exclude="*_schema.json" ${patterns_src}/{constants,encodings,includes,magic,patterns} "$out/Applications/imhex.app/Contents/MacOS"
```
```suggestion
    rsync -av --exclude="*_schema.json" ${patterns_src}/{constants,encodings,includes,magic,nodes,patterns} "$out/Applications/imhex.app/Contents/MacOS"
```

viraptor reviewed (commented):
`nodes` was missed in the linux non-macos package as well. It comes from the patterns repo.

afh commented:
The order of work/commits that has worked well for me for similar PRs is:

* version bump
* add support for darwin
* refactor & modernize

All in one PR. Th order of commits allows to split work/commits into separate PRs if the reviewers cannot decide "quickly" on a path forward. This has rarely happened in my experience.

emilazy commented:
I would usually refactor (and/or reformat) before significant changes to add platform support, to avoid doing work just to redo it, so to speak. (Rule of thumb: If there are a ton of lines in your commit diffs that don’t show up in the final PR diff, you’ve probably wasted time and not really made it any easier on reviewers.)

But otherwise, agreed with @afh.

pcasaretto committed:
imhex: add support for darwin

Co-authored-by: Anderson Torres <torres.anderson.85@protonmail.com>

pcasaretto commented:
Ready for review! 
@kashw2 I'll glady add myself as a mantainer to help with darwin if you want.

khaneliman commented:
> Ready for review! 
> 
> @kashw2 I'll glady add myself as a mantainer to help with darwin if you want. 

I'd recommend it, it's nice having a Darwin maintainer on most derivations that require some legwork like that.

cafkafk reviewed (changes requested):
Thanks for working on this :)

#323501 just landed, meaning this will have to be rebased on top of it.

I don't think we should have a version update in this PR, it should focus purely on the darwin support, so it doesn't block updating the package.

Refactoring code that's contributed in the same PR should not be it's own separate commit. Other refactoring of the code should ideally also go in it's own commit, as it makes it easier to review either part separately.

afh commented:
Thanks for chiming in on this @cafkafk, can you point me to a documented guideline or recommendation, e.g. in a README, regarding refactoring changes and commits in nixpkgs PRs? In my contributions I've come across what seemed different opinions and it'd be helpful to have a reference to check in the future and quote in reviews.

viraptor commented:
@afh https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md - although some points are vague and you'll find people having various interpretations. (Worst case, if something's not written down and you have other opinion, don't be afraid to push back a bit)

afh commented:
Thanks I'll go over the document again and re-read it, some of its aspects may have eluded me the first time around 🙂

pcasaretto commented:
I've rebased and kept only the changes relevant for adding support for darwin (dropped the version bump).
The build is now failing with 

```
calling 'postFixup' function hook 'autoPatchelfPostFixup'                                                                                             
Traceback (most recent call last):                                                                                                                    
  File "/nix/store/yxvjvrvki76nf4bshfyccf03jwv41zgp-auto-patchelf.py", line 414, in <module>                                                          
    with open_elf(interpreter_path) as interpreter:
  File "/nix/store/1sgajx2r3bkriyxzwsahhva63p08pmac-python3-3.12.4/lib/python3.12/contextlib.py", line 137, in __enter__
    return next(self.gen)
           ^^^^^^^^^^^^^^
  File "/nix/store/yxvjvrvki76nf4bshfyccf03jwv41zgp-auto-patchelf.py", line 27, in open_elf
    yield ELFFile(stream)
          ^^^^^^^^^^^^^^^
  File "/nix/store/va7nr2xsgry9vf2x9mdqwskgyqf86fr4-python3-3.12.4-env/lib/python3.12/site-packages/elftools/elf/elffile.py", line 84, in __init__
    self._identify_file()
  File "/nix/store/va7nr2xsgry9vf2x9mdqwskgyqf86fr4-python3-3.12.4-env/lib/python3.12/site-packages/elftools/elf/elffile.py", line 570, in _identify_f
ile
    elf_assert(magic == b'\x7fELF', 'Magic number does not match')
  File "/nix/store/va7nr2xsgry9vf2x9mdqwskgyqf86fr4-python3-3.12.4-env/lib/python3.12/site-packages/elftools/common/utils.py", line 80, in elf_assert
    _assert_with_exception(cond, msg, ELFError)
  File "/nix/store/va7nr2xsgry9vf2x9mdqwskgyqf86fr4-python3-3.12.4-env/lib/python3.12/site-packages/elftools/common/utils.py", line 143, in _assert_wi
th_exception
    raise exception_type(msg)
elftools.common.exceptions.ELFError: Magic number does not match
```

Is this recent code relevant for darwin?

afh commented:
Given that ELF is the binary format for Linux executables and Mach-o is the binary format for darwin executables I'd say no and guarding these new additions with `stdenv.isLinux` should be alright.

pcasaretto commented:
Done! However, I'm feeling the need to refactor the code as there are too many conditional expressions.
I've seen a couple of packages use callPackage to split the code in more files.
I was thinking of taking a different approach and keep all the common attributes in the main file and then deep merging the attributes of a OS specific secondary file.
Darwin for example would add `postPatch`, `makeWrapper` to `nativeBuildInputs`, etc.
While linux would add `autoPatchelfHook`, its related attributes and customization of `postInstall`.
WDYT?

afh commented on pkgs/by-name/im/imhex/package.nix:
```diff
@@ -73,31 +86,48 @@ stdenv.mkDerivation rec {
   # however, we will append to RUNPATH ourselves
-  autoPatchelfIgnoreMissingDeps = [ "*.hexpluglib" ];
-  appendRunpaths = [
+  autoPatchelfIgnoreMissingDeps = lib.lists.optional stdenv.isLinux [ "*.hexpluglib" ];
+  appendRunpaths = lib.lists.optional stdenv.isLinux [
```
I haven't come across `lib.ists.optional` in my nixpkgs work so far and will need to read up on it. Mind providing some context as to why this was used here instead of just `lib.optionals`, @pcasaretto? 🙂

pcasaretto replied:
It was incorrect. Thanks for catching that!`optional` would have inserted the list inside another list.
However, I fail to understand how `optional` and `optionals` are accessible in the root `lib`.
From the documentation they seem to belong to the `lists`  subattribute.

fidgetingbits replied:
A lot of the functions are aliased into lib. You can use noogle.dev to check if a function is aliased, for example see the bottom list of aliases for optional:

https://noogle.dev/f/lib/lists/optional

pcasaretto replied:
Bookmarked!

afh commented:
Interesting idea, @pcasaretto, it's a pattern I haven't seen much, if at all. In order to get a better understanding I'd like to see a more complete example, if that isn't too much to ask. Another approach could be to look for other packages that tackle the different needs to compile a package on Linux and Darwin and see what they have done.
ℹ️ I don't have as much time for nixpkgs this week as I usually do, feel free to move forward with this if others can provide the feedback you find helpful. 🙂

pcasaretto commented:
Applied automatic formatting into a separate commit to facilitate the review. Let me know if I should have amended into one single commit.

afh commented:
I'm okay with the re-formatting changes to be an extra commit, thanks :)

viraptor reviewed (approved):
I don't think we need to bikeshed how to improve this package here. (unless you want to) The darwin fixes work fine, which was the purpose. We can always do another PR with a cleanup / refactor to something nicer.
The current code doesn't go past the necessary complexity with a few conditionals, so 👍

pcasaretto commented:
Whenever you can @cafkafk, this is ready for your review.

viraptor commented:
Just noticed there's some issue with intel darwin:
```
[ 70%] Building CXX object plugins/builtin/CMakeFiles/builtin.dir/source/content/providers/process_memory_provider.cpp.o
/tmp/nix-build-imhex-1.35.3.drv-0/ImHex-source-1.35.3/plugins/builtin/source/content/providers/process_memory_provider.cpp:12:14: fatal error: 'mach/arm/kern_return.h' file not found
   12 |     #include <mach/arm/kern_return.h>
      |              ^~~~~~~~~~~~~~~~~~~~~~~~
```

pcasaretto commented:
Using `nix-build -A imhex --system x86_64-darwin` I'm able to replicate locally.
Thanks for the heads up viraptor!
I'm on it.

pcasaretto commented:
After using `overrideSDK` to make x86 darwin use the correct version (11) of the SDK, I encountered errors mentioning duplicated typedefs.

```
In file included from /tmp/nix-build-imhex-1.35.3.drv-0/ImHex-source-1.35.3/plugins/builtin/source/content/providers/process_memory_provider.cpp:13:  
In file included from /nix/store/2l9s8m6hb82d4qghp9bsqd65radv11q1-libSystem-11.0.0/include/mach/arm/vm_types.h:72:                                    
/nix/store/2l9s8m6hb82d4qghp9bsqd65radv11q1-libSystem-11.0.0/include/arm/_types.h:55:3: error: typedef redefinition with different types ('union __mbs
tate_t' vs 'union __mbstate_t')                                                                                                                       
   55 | } __mbstate_t;                                                                                                                                
      |   ^                                                                                                                                           
/nix/store/2l9s8m6hb82d4qghp9bsqd65radv11q1-libSystem-11.0.0/include/i386/_types.h:79:3: note: previous definition is here                            
   79 | } __mbstate_t;
      |   ^
```

I tracked down the import to this file https://github.com/WerWolv/ImHex/blob/de8465a8f4f4945d1c87a499872749b65e522e2d/plugins/builtin/source/content/providers/process_memory_provider.cpp#L13.
It seemed strange that the code was importing an architecture-specific header, so I patched it and voila, it compiled.

I'm taking this finding to WerWolv to check, but in the meanwhile, should we proceed with the patch?

viraptor commented:
Alternatively could raise that as a PR upstream and `fetchpatch` in the build, but I think the current way is ok too - they're responsive, so I hope they'll fix it soon and we can drop it in the next version.

AndersonTorres commented on pkgs/by-name/im/imhex/package.nix:
```diff
@@ -107,6 +148,6 @@ stdenv.mkDerivation rec {
       cafkafk
     ];
-    platforms = platforms.linux;
+    platforms = platforms.unix;
```
Since other platforms are ruled out by your throw [here](https://github.com/NixOS/nixpkgs/pull/330303/files#diff-cb182ce59baa3510525615b5be3ccdca218b8a32b330b315bfd98a29374c04bdR139),

```suggestion
    platforms = platforms.linux ++ platforms.darwin;
```

is more accurate

siraben commented:
Result of `nixpkgs-review pr 330303` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>imhex</li>
  </ul>
</details>

pcasaretto commented:
Hey everyone. I've rebased main, squashed the commits down and used the patch strategy suggested by @viraptor .
Given everything returns with a green status, should we merge?

siraben commented:
Result of `nixpkgs-review pr 330303` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>imhex</li>
  </ul>
</details>

mirkolenz commented:
Result of `nixpkgs-review pr 330303` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>imhex</li>
  </ul>
</details>

afh commented on pkgs/by-name/im/imhex/package.nix:
```diff
@@ -46,20 +59,34 @@ stdenv.mkDerivation rec {
-    pkg-config
-    rsync
+  patches = [
+    (fetchpatch {
```
A commit why this patch is necessary and whether it can be removed with the next or future update would be helpful.

afh commented on pkgs/by-name/im/imhex/package.nix:
```diff
@@ -46,20 +59,34 @@ stdenv.mkDerivation rec {
   ];
 
+  postPatch = lib.optionalString stdenv.isDarwin ''
+    substituteInPlace cmake/modules/PostprocessBundle.cmake \
```
Providing context why this is necessary would be helpful too

afh reviewed (approved):
Thanks for putting in all the time and effort into this, @pcasaretto, much appreciated 🙏

Please find below two minor, optional suggestions for improvement.

pcasaretto commented:
Done @afh! The comments will surely help.
Given that we are building successfully for x86-64 ( ❤️ @mirkolenz ), I'm going to work with the maintainer to get that PR merged.

pcasaretto commented:
Hey @cafkafk , can you review this again whenever you get a chance ?

NeQuissimus reviewed (approved):
LGTM

cafkafk reviewed (approved):
Well done surviving such a long PR process, and thanks for working so actively to fix this! This looks good enough to me, and I think we should just get it in at this point. Again, thanks for all the time put into this!
